From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ProdPreva1l <prod.preva1l@gmail.com>
Date: Fri, 26 Apr 2024 05:41:51 +1000
Subject: [PATCH] Just modify the header


diff --git a/src/main/java/net/dxtrus/DxtrusConfig.java b/src/main/java/net/dxtrus/DxtrusConfig.java
index 7fb6477310d9c621eac3ba00e6693f372c49ad04..3268c15488627ec3e55a466fcbde25e852d37429 100644
--- a/src/main/java/net/dxtrus/DxtrusConfig.java
+++ b/src/main/java/net/dxtrus/DxtrusConfig.java
@@ -15,11 +15,12 @@ import java.util.List;
 import java.util.Map;
 import java.util.logging.Level;
 
+@SuppressWarnings("unused")
 public class DxtrusConfig {
     private static final String HEADER =
             """
-                This is the main configuration file for Lamington (Dxtrus fork of Purpur)
-                Not much to configure but it will grow!
+            This is the main configuration file for Lamington (Dxtrus fork of Purpur)
+            Not much to configure but it will grow!
             """;
     private static File CONFIG_FILE;
     public static YamlConfiguration config;
